lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -811,6 +811,65 @@ class DecomposeAtenTOp : public OpRewritePattern<AtenTOp> {
 };
 } // namespace
 
+// Decompose `aten.stack` into `aten.unsqueeze` and `aten.cat`.
+namespace {
+class DecomposeAtenStackOp : public OpRewritePattern<AtenStackOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenStackOp op,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Value> tensors;
+    if (!getListConstructElements(op.getTensors(), tensors)) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: the tensor list is not from list construct");
+    } else if (tensors.empty()) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: the tensor list is empty");
+    }
+
```

Comment:
The code towards the end makes the assumption that all the tensors have the same type. Can you add a check here to make sure this is true and return early if it's not?

Reply:
Do we need to support tensors of different dtypes in the input list? PyTorch actually supports type promotion on all input tensors of `cat` and `stack` ops.

Reply:
This case can be left as `unimplemented: ...`. If you do implement this, you just need to be careful when creating the `!torch.list` type for the operand that will be passed to `cat`. If the dtypes are different for the tensors in the list, then the contained type in `!torch.list` should be a type with no `dtype`

Reply:
I do support this. Added a new e2e case `TensorsStackPromoteDTypeModule_basic` to test type promotion.